### ISSUE TITLE
Difference between FairScale and FlexFlow

### ISSUE THREAD

**Issue by mattiasmar** [facebookresearch/fairscale#665]
## ❓ Questions and Help
What are the similarities and differences between Fairscale and [FlexFlow](https://github.com/facebookresearch/dlrm-flexflow)?

**Comment by min-xu-ai** [comment]
@msbaines

**Comment by Vittorio-Caggiano** [comment]
Hi @mattiasmar, FairScale and FlexFlow are conceptually different both in the design and in the use cases they address. FairScale is a library where you can compose different methods for scaling and making more efficient training of neural networks. FlexFlow is a framework with the ability to search for parallelization strategies across the sample, operator, attribute, and parameter dimensions for a specific underlying architecture. FairScale doesn't offer those features.

**Comment by mattiasmar** [comment]
Thanks @Vittorio-Caggiano. I understand that FairScale enables scaling by allowing parallelization. Which out of the SOAP dimensions does FairScale support parallelization for (support in the sense that the user provides the amount and type of parallelization and the library cares for the actual splitting)?

**Comment by Vittorio-Caggiano** [comment]
FairScale is a library of components where you need to choose what kind of parallelism and/or optimization works best for you. There is no option for automatic analysisinp your model and your infra to determine that for you.

**Comment by min-xu-ai** [comment]
Hope this question is answered.

**Comment by mattiasmar** [comment]
@Vittorio-Caggiano Thanks again for your answer.
From your answers I understand that FairScale  and FlexFlow have  equivalent components to implement different kinds parallelism and to measure inference and training latencies under a specified parallelism (that can be unique for each layer).
I further understand that the essential difference between the packages lies in that FlexFlow besides the parallelism enablement also has mechanics to search for an efficient parallelism strategy (strategy= a per layer mapping of parallelism).
Do you think this is an accurate summary? If yes, then I want to ask; If I am looking for an implementation that enables parallelism, what are the reasons to choose to work with one system over the other?

**Comment by Vittorio-Caggiano** [comment]
@mattiasmar sorry for the late reply, I didn't see this msg. The main difference in using FairScale vs. FlexFlow is the difference to use a library vs a Framework. FairScale is a library: you can import the library and with small additions to your code (if you use PyTorch DDP) you should be able to use the supported parallelism and easily extend / customize them (see the examples in the ReadMe). FlexFlow will need some more configurations and changes to your code to fully take advantage of the framework. The major advantages of FairScale are the easy set up (just a pip install and a few changes to load the APIs), composition (you can combine different APIs) and customization (the code should be pretty easy to follow if you have read the paper). Still, this requires a bit more understanding of the parallelism options. I am not an expert on FlexFlow though. I would suggest to reach out @jiazhihao and the FlexFlow team for more details.

**Comment by mattiasmar** [comment]
Thanks @Vittorio-Caggiano
